### ISSUE TITLE
Resources data being lost when reading from info.resource returned by _on_item_loaded

### ISSUE THREAD

**Issue by Jack-023** [derkork/godot-resource-groups#13]
It was a bit hard to describe the issue in a sentence so sorry about the title. I'll describe the scenario I have and what I am seeing (vs expectation).

I have a class.
```gdscript
class_name Item
extends Resource

@export var name: String
@export var icon: Texture
@export var tooltip: String
...
```

It serves as a container that holds various game objects that can be stored in an inventory. These objects can be very different in their behaviour, all they have in common is that they get stored in the same array of `Item`. Each type of item has it's own "item type", for example:
```gdscript
class_name SpellModItem
extends Item

@export var mod: SpellMod
```
Where a mod is an item that takes a spell and changes some property of it.
```gdscript
class_name SpellMod
extends Resource

#virtual
func mod(spell: Spell) -> void:
    pass
```

There are different categories of mod that share the same behaviour, they extend `SpellMod` and override the `mod` method.
```gdscript
class_name FlightPathMod
extends SpellMod

# These values are from the Spell type, it samples the curve to get its new relative rotation value each frame.
@export var rotation_curve: Curve
@export var period_length: int

func mod(spell: Spell) -> void:
    spell.rotation_curve = rotation_curve
    spell.period_length = period_length
```

This is a bit simplified compared to the implementation but is enough to illustrate what I am trying to do. I have an autoload called `ItemDb` that loads all data in the background when the game starts. It uses a resource group of `Item` resource files to load all my games items into a Dictionary of `Item:name -> Item`. I then expose a method on this `ItemDb` singleton to retrieve items by their names.

I can then load the inventory like so:
```gdscript
player_inventory.append(ItemDb.get_by_name("wave flight") as SpellModItem)
player_inventory.append(ItemDb.get_by_name("zig zag flight") as SpellModItem)
```
and access the inner `FlightPathMod`s when I need to actually apply them.

The actual `.tres` files the ResourceGroup loads are the `Item` subtypes (eg `SpellModItem`). In the spell mod case, the `spell_mod` property is set in the `.tres` file as the path to another `.tres` representing an instance of a `SpellMod` subtype (eg `FlightPathMod`). A psudo-text? representation of the resource files would be something like.
```
# wave_flight_item.tres
spell_mod = "res://.../wave_flight_mod.tres"
-----
# zig_zag_flight_item.tres
spell_mod = "res://.../zig_zag_flight_mod.tres"
```

Sorry that was a bit lengthy. 

The actual issue I am seeing is that the `spell_mod` property of all `SpellModItem` who's `spell_mod` is of type `FlightPathMod` is being set to the value of the first `FlightPathMod` that is loaded, after it has been stored outside of the `info.resource` property given to `_on_item_loaded`.

So wave flight mod item loads first 
I hit a breakpoint in the `_on_item_loaded` callback, the data in `info.resource` is correct.
I store the data in a dictionary `_item_db[info.resource.name] = info.resource`  the data stored in the dictionary is also correct.
zig zag flight mod item loads second
I hit a breakpoint in the `_on_item_loaded` callback, the data in `info.resource` is correct.
I store the data in a dictionary `_item_db[info.resource.name] = info.resource` the data stored in the dictionary is **incorrect**

It looks like 
```
{
    name: "zig zag flight", # correct
    icon: { ... }, # correct
    tooltip: "Lorem Ipsum", # correct
    spell_mod: { # wave flight resource data including incorrect path } # incorrect
}
```

At this point, the data is still correct when inspecting the value in `info.resource` so it hasn't been dereferenced. I assumed this was an issue with Godot dictionaries, so I created a minimal repro project to demonstrate the issue (using  ResourceLoader directly instead of ResourceGroups) but it was working as expected.

I am pretty confused as to what is going on here and it took a while to even find where the data was being lost. If you need any more info to help troubleshoot this let me know.

**Comment by derkork** [comment]
Having a simple reproducer would really help getting to the bottom of this. In general the resource group plugin doesn't modify any of the resources it loads, it just call's Godots APIs to load the resources (e.g. `ResourceLoader.load`) and then ships the resources out to the caller. Can you try if this works if you manually load the resource (e.g. just `load("res://.../SomeSpellModItem.tres")`). 

Also check if you have code that modifies the resources linked in the Spell mod item. Remember that resources are shared, so if you link the same resource to multiple spell mod items and change it, it will change in all spell mod items because all of them share the same instance.

**Comment by Jack-023** [comment]
Thanks for having a look at this. I have spent a bit more time trying to troubleshoot this issue and I am pretty confident that is an issue with Godot, not the plugin. Sorry for the false alarm.

I tried recreating this issue by changing my initial MRP I set up to report this as an engine bug (where I couldn't reproduce it) on a different branch that used the plugin. I was also unable to reproduce the issue in the new project when using the addon.

I went back to my main project, and saw and issue with the resources without even running the game. When I open `wave_path_mod_item.tres` I can see in the inspector that the spell mod property is showing `zig_zag_flight_mod.tres`. When I look at `wave_path_mod_item.tres` in a text editor, the actual file contents are 
```tres
[ext_resource type="Resource" uid="uid://deadbeef" path="res://resources/spell_mods/flight_path/wave/wave_flight_mod.tres" id="0_foobar"]
...

spell_mod = ExtResource("0_foobar")
```
which is what I actually want and was expecting. I thought I was seeing this last week but at the time I wrote up the issue here, it looked right up until the point I described initially in this post (when moving the data in memory at runtime). It looks like there is some issue with state that is contributing to this and making it difficult to reproduce.

I am guessing that there is something wrong with how values are being cached / custom classes are being loaded by Godot. I tried deleting my `.godot` directory but the issue still exists.
I'll close this issue and make an issue on the Godot repo and link it here in case anybody else is unfortunate enough to find themselves running into this problem.
